### PR TITLE
OEUI-238: Toggle lab urgency from shared draft

### PR DESCRIPTION
**Pull request by IAMOTZ**
### **JIRA TICKET NAME**
[OEUI-238: User should be able to toggle the urgency of a lab order from the draft.](https://issues.openmrs.org/browse/OEUI-238)

### **Summary**
This pull request adds the functionality that enables users to toggle the urgency of Lab Orders from the new shared draft.

### CODE REVIEW

File: app/js/actions/draftActions.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+  const orderUuid = order.uuid;
+  let orderUrgency;
+  const hasUrgencyProperty = Object.prototype.hasOwnProperty.call(order, 'urgency');
+  if (hasUrgencyProperty && order.urgency === constants.ROUTINE) {
+    orderUrgency = constants.STAT;
+  } else if (hasUrgencyProperty && order.urgency === constants.STAT) {
+    orderUrgency = constants.ROUTINE;
+  } else {
+    orderUrgency = constants.STAT;
+  }
```
Comment (Efosaok) [inline app/js/actions/draftActions.js]: since you are basically toggling between two values, I think the conditional blocks you are using is a bit of an overkill. please refactor to something like this.

```
let orderUrgency = constants.STAT
if (hasUrgency) {
   orderUrgency = constants[order.urgency]
}
```

File: app/js/components/orderEntry/OrderEntryPage.jsx [hunk 2]
```diff
@@ -14,6 +14,7 @@ import getDateFormat from '../../actions/dateFormat';
 import activeOrderAction from '../../actions/activeOrderAction';
 import { fetchPatientRecord, fetchPatientNote } from '../../actions/patient';
 import { setSelectedOrder } from '../../actions/orderAction';
+import toggleDraftLabOrderUrgency from '../../actions/draftActions';
```
Comment (Efosaok) [inline app/js/components/orderEntry/OrderEntryPage.jsx]: `name of import` should match `export-default`

File: tests/actions/draftActions.test.js [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        TOGGLE_DRAFT_LAB_ORDER_URGENCY,
+        order: {
+          ...labOrder, 
+          urgency: 'ROUTINE',
+        },
+      };
+      const store = mockStore({});
+      await store.dispatch(toggleDraftLabOrderUrgency(labOrder), () => {
+        expect(store.getActions()).toEqual(expectedAction);
+      });
```
Comment (Efosaok) [inline tests/actions/draftActions.test.js]: `store.dispatch` is a `Promise` based function and to the best of my knowledge, doesn't take in a second parameter. since you are using the `await` keyword, refactor this to

```
await store.dispatch(toggleDraftLabOrderUrgency(labOrder));
expect(store.getActions()).toEqual(expectedAction);
 ```

File: tests/actions/draftActions.test.js [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
```
Comment (Efosaok) [inline tests/actions/draftActions.test.js]: refer to above comment

File: tests/actions/draftActions.test.js [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+        TOGGLE_DRAFT_LAB_ORDER_URGENCY,
+        order: {
+          ...labOrder, 
+          urgency: 'STAT',
+        },
+      };
+      const store = mockStore({});
+      await store.dispatch(toggleDraftLabOrderUrgency(labOrder), () => {
+        expect(store.getActions()).toEqual(expectedAction);
+      });
```
Comment (Efosaok) [inline tests/actions/draftActions.test.js]: same here

File: app/js/actions/draftActions.js (hunk 1 above)
Comment (IAMOTZ) [reply app/js/actions/draftActions.js]: Hi @Efosaok 
Yes, you are right, I am toggling between two values, however, the next value actually depends on the previous and that is why the series of checks is needed.
Let me try and explain the checks in plain English:
- If the order has urgency property and its previous value is ROUTINE, the urgency should become STAT.
- If the order has urgency property and its previous value is STAT, the urgency should become ROUTINE.
- If the order does not have urgency property, the urgency should become STAT.

I hope this clarifies your concern.

File: app/js/components/orderEntry/OrderEntryPage.jsx (hunk 2 above)
Comment (IAMOTZ) [reply app/js/components/orderEntry/OrderEntryPage.jsx]: The file `draftAction` is still going to contain a lot of actions that belong to the draft some of which are to be implemented in [this](https://issues.openmrs.org/browse/OEUI-236) and [this](https://issues.openmrs.org/browse/OEUI-239) tickets.
That is the reason why I named the file `draftActions`, once other actions are added to the file, the `export-default` won't be needed anymore.

File: app/js/actions/draftActions.js (hunk 1 above)
Comment (Efosaok) [reply app/js/actions/draftActions.js]: I understand exactly what you are trying to do. but I think what I missed out is that you should set the `constant` variable to this.

```
export const STAT = 'ROUTINE'
export const ROUTINE = 'STAT'
```
and then just use the code up there to replace what you have already. this would make your  code more concise..

File: app/js/components/orderEntry/OrderEntryPage.jsx (hunk 2 above)
Comment (Efosaok) [reply app/js/components/orderEntry/OrderEntryPage.jsx]: If there was a rule to lint this, you know you'll probably have to fix it. I understand your point but the code written at any point in time should adhere to conventions used at that point in time.

File: app/js/actions/draftActions.js (hunk 1 above)
Comment (IAMOTZ) [reply app/js/actions/draftActions.js]: Hmm..., now I get what you are saying.
I have updated the code with an implementation that reflects what you said above.
Thanks man, I was able to reduce some line of codes(and thinking effort) over there.

File: tests/actions/draftActions.test.js (hunk 4 above)
Comment (IAMOTZ) [reply tests/actions/draftActions.test.js]: Thanks.
The code has been updated!

File: app/js/components/orderEntry/OrderEntryPage.jsx (hunk 2 above)
Comment (Efosaok) [reply app/js/components/orderEntry/OrderEntryPage.jsx]: since this PR and the related PR's will be merges soonest, you can leave it as it is but just bear in mind that this is not acceptable under normal circumstances.
